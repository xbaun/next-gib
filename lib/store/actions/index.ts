export * from './auth/auth.action';
export * from './search/search.action';
export * from './issue/issue.action';
