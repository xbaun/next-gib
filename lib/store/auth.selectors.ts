import { RootState } from './index';

export const getAuth = (state: RootState) => state.auth;
export const getAuthToken = (state: RootState) => state.auth.token;
