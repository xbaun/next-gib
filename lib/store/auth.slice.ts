import { createAsyncThunk, createSlice, PayloadAction } from "@reduxjs/toolkit";
import fetch from "isomorphic-unfetch";

type InitialState = {
    token: string | undefined;
    error: boolean;
    authenticating: boolean;
};

export const initialState: InitialState = {
    token: undefined,
    error: false,
    authenticating: false,
};

export const fetchAccessToken = createAsyncThunk<{ data?: { token: string }, error: boolean }, { code: string }>(
    "auth/fetchAccessToken",
    async ({ code }) => {
        console.log("login.fulfilled", "start");

        const url = new URL(`/api/auth/github`, location.href);
        url.searchParams.append("code", code);

        const data = await fetch(url.toString()).then((res) => {
            if (res.ok) {
                return res.json();
            } else {
                return undefined;
            }
        });

        if (data && "accessToken" in data && data.accessToken) {
            return { error: false, data: { token: data.accessToken } };
        } else {
            return { error: true };
        }


    }
);

export const authSlice = createSlice({
    name: "auth",
    initialState,
    reducers: {
    },
    extraReducers: (builder) => {
        builder
            .addCase(fetchAccessToken.fulfilled, (state, action) => {
                console.log("login.fulfilled", action);
                return { ...state, authenticating: false, error: action.payload.error, token: action.payload.data?.token };
            })
            .addCase(fetchAccessToken.pending, (state, action) => {
                console.log("login.pending", action);
                return { ...state, authenticating: true, error: false };
            });
    },
});

export const { } = authSlice.actions;

export default authSlice.reducer;
