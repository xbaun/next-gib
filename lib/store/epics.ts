import { createAction } from '@reduxjs/toolkit';
import { ActionsObservable, combineEpics, ofType } from 'redux-observable';
import { map } from 'rxjs/operators';
import { fetchReposBySearchTerm, SetSearchRepoTerm, setSearchRepoTerm } from './search-repo.slice';
//
//
// export const fetchUserAccessToken = createAction<{ code: string }>('fetchUserAccessToken');
// export type  FetchUserAccessToken = ReturnType<typeof fetchUserAccessToken>;
//
//
// const epic = (action$: ActionsObservable<FetchUserAccessToken>) => action$.pipe(
//     ofType<FetchUserAccessToken>(fetchUserAccessToken.type),
//     map((action) => fetchUserAccessToken({ code: ""}))
// );
//




const fetchReposBySearchTermEpic = (action$: ActionsObservable<SetSearchRepoTerm>) => action$.pipe(
    ofType<SetSearchRepoTerm>(setSearchRepoTerm.type),
    map((action) => fetchReposBySearchTerm({ term: action.payload.term}))
);


export const rootEpic = combineEpics(
    fetchReposBySearchTermEpic
);
