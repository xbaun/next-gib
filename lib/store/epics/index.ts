import { combineEpics } from 'redux-observable';
import authEpics from './auth.epic';
import commentEpics from './comment.epic';
import issuesEpic from './issues.epic';
import searchEpics from './search.epic';

export default combineEpics(authEpics, searchEpics, issuesEpic, commentEpics);
