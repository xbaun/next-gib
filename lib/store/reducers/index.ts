import { combineReducers } from 'redux';
import { StateType } from 'typesafe-actions';
import authReducer from './auth.reducer';
import { commentsReducer } from './comments/comments.reducer';
import { issuesReducer } from './issues/issues.reducer';
import { searchReducer } from './search/search.reducer';

const rootReducer = combineReducers({
    auth: authReducer,
    search: searchReducer,
    issues: issuesReducer,
    issueComments: commentsReducer
});

export type RootStateType = StateType<typeof rootReducer>;

export default rootReducer;
