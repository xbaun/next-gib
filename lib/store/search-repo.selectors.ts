import { RootState } from './index';

export const getSearchReopState = (state: RootState) => state.searchRepo;
