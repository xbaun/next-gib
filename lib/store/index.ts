import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import { createEpicMiddleware } from 'redux-observable';
import authReducer from './auth.slice'
import { rootEpic } from './epics';
import searchRepoReducer from './search-repo.slice'

const epicMiddleware = createEpicMiddleware();

const store =  configureStore({
    reducer: {
        auth: authReducer,
        searchRepo: searchRepoReducer
    },
    middleware: getDefaultMiddleware().concat([epicMiddleware]),
    devTools: true
});

epicMiddleware.run(rootEpic as any);

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch

export type AppThunkAPI =  {
    dispatch: AppDispatch
    state: RootState
}

export default store;



