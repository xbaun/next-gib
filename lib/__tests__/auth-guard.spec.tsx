import { render } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import AuthGuard from '../components/auth-guard';
import { store } from '../store';

describe('AuthGuard', () => {
    const useRouter = jest.spyOn(require('next/router'), 'useRouter');
    let push: jest.Mock;

    beforeAll(() => {
        push = jest.fn();

        useRouter.mockClear();
        useRouter.mockImplementation(() => ({
            route: '/',
            push
        }));
    });

    it('should redirect correctly to route: /auth', async () => {
        const { container } = render(
            <Provider store={store}>
                <AuthGuard fallback={'/'} redirect={'/auth'}>
                    <></>
                </AuthGuard>
            </Provider>
        );

        expect(push.mock.calls[0][0]).toBe('/auth');
        expect(container).toMatchSnapshot();
    });
});
