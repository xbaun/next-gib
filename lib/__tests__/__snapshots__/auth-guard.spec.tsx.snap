// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthGuard should redirect correctly on route: /auth 1`] = `
<div>
  ...
</div>
`;

exports[`Should render correctly on route: /users/nikita 1`] = `
<div>
  ...
</div>
`;
